src/compat/eme/close_session.ts

```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright 2015 CANAL+ Group
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Observable,
+  of as observableOf,
+  race as observableRace,
+  timer as observableTimer,
+} from "rxjs";
+import {
+  catchError,
+  map,
+  mergeMap,
+} from "rxjs/operators";
+import castToObservable from "../../utils/cast_to_observable";
+import { isFirefox } from "../browser_detection";
+import { ICustomMediaKeySession } from "./custom_media_keys";
+
+/**
+ * Close session and returns and observable that emits when
+ * the session is closed.
+ * @param {MediaKeySession|Object} session
+ * @returns {Observable}
+ */
+export default function closeSession$(
+  session: MediaKeySession|ICustomMediaKeySession
+): Observable<unknown> {
+  return observableRace(
+    castToObservable(session.close()),
+    // If the session is not closed after 1000ms, try
+    // to call another method on session to guess if
+    // session is closed or not.
+    observableTimer(1000).pipe(
+      mergeMap(() => {
+        const tryToUpdateSession$ = castToObservable(
+          session.update(new Uint8Array(1))
+        );
+        return tryToUpdateSession$.pipe(
+          // Update has resolved, so we can't know if session is closed
+          map(() => { throw new Error("Compat: Couldn't know if session is " +
+                                      "closed"); }),
+          catchError((err) => {
+            if (isFirefox) {
+              // On Firefox, after that `update` has thrown,
+              // the `closed` Promise on session resolves.
+              const sessionIsClosed$ = castToObservable(session.closed);
+              return observableRace(
+                sessionIsClosed$,
+                observableTimer(1000).pipe(
+                  map(() => { throw new Error("Compat: Couldn't know if session is " +
+                                              "closed"); })
+                )
+              );
+            }
+            // The caught error can tell if session is closed.
+            if (err.message === "The session is already closed.") {
+              return observableOf(null);
+            }
+            throw new Error("Compat: Couldn't know if session is closed");
+          })
+        );
+      })
+    )
+  );
```

Comment:
All this code looks still too specific. We might break at any widevine update and decide to throw.

Reply:
Well, we decide to throw if none of known hints that session is closed is found. It means that we may throw if, for a reason or another, no hint is found (Widevine update or not).

I try to remove browser check.

Reply:
What is the impact of throwing here?
Is it risky to throw here even if we have a high chance to have succeeded?

Reply:
The problem is that throwing might completely stop playback, which I do not think is necessary here.
So maybe throw here, but I think we should better handle that failure elsewhere.

Reply:
Ok. The compat close session still throws, and i've handled the error downstream.